### ISSUE TITLE
Single line text entry glitch when newline entered

### ISSUE THREAD

**Issue by iacore** [david-vanderson/dvui#35]
here you can see I type `aa<Enter>b<Enter><Enter><Enter><Enter>`

https://github.com/david-vanderson/dvui/assets/74560659/4a5975a2-3172-431b-a191-380f250d404e

**Comment by david-vanderson** [comment]
Thanks very much for the video!  You are right that is not good.

My initial idea for this was to try to have the "no newlines" logic outside the TextEntryWidget, on the general theme that it's already complicated enough.  But my "delete the newlines after the fact" strategy causes this temporary visual glitch.

I think I have a better idea for how to intercept the events before they get to the widget.  Let me try again and we can see how it works out.

**Comment by iacore** [comment]
Most importantly, pressing Enter key on single-line textentry should be some kind of event.

**Comment by david-vanderson** [comment]
That is an excellent point.  There's nothing in dvui so far to handle this kind of thing but I have thought about it somewhat.  In other UI frameworks this kind of thing is wrapped up in "dialog" handling.  I'm very unsure what the right thing to do here is, so I'm waiting until we get some experience with actually trying to do it first.

I'll try to add something to the example along these lines.  Thanks!

**Comment by iacore** [comment]
I met this problem today [here](https://github.com/iacore/janet-graphical-repl).

My workaround:

Detect if the text ends in \n. If yes, use the data.

**Comment by david-vanderson** [comment]
I've committed some changes to TextEntryWidget because my previous ideas for how to intercept events didn't pan out.  Hopefully this new approach is better.

I updated the single-line example to show how to intercept events (like pressing enter).  Can you take a look and let me know what you think?

**Comment by iacore** [comment]
I assume you mean [this change](https://github.com/david-vanderson/dvui/commit/d5b921b68b5721b7e423fffbc08299e19c9f18db#diff-3652b591c46fb901efc6880ffab840b7c40d451d6e2eda61001ee6b492e77920R324-R365).

This requires the user to know the internals of dvui.

I imagined the api to be like this:

```zig
const pressed: bool = te.sense_key(.enter, .{.drop_event=true});
```

**Comment by iacore** [comment]
I have found a workaround to the glitch effect: `.scroll_vertical_bar = .hide`.

**Comment by david-vanderson** [comment]
Yes that's the change - you're right that accessing the events like that isn't very slick.  That egui api might work well here, but I'm worried about designing something without enough examples.  It sounds weird, but this is the first example of wanting to override a widget's event handling.

**Comment by iacore** [comment]
If you want examples, think of an application with shortcuts everywhere. Global shortcuts and per-widget shortcuts. Also widgets with shortcuts nested within widgets with shortcuts.

**Comment by david-vanderson** [comment]
I've committed some more changes to TextEntryWidget, making single-line the default.  Thanks for the push there!

For keyboard shortcuts, I ran into a snag.  Because dvui processes multiple events each frame, the widget could get focus and receive the shortcut in the same frame.  That means overriding normal event processing needs to happen inside the widget's event loop.

I think the results are pretty good.  It's not super slick, but it's straightforward and makes it obvious how to do more complicated things.  Example:
```zig
var te = dvui.TextEntryWidget.init(@src(), .{ .text = &text_entry_buf }, .{});
try te.install();

const emo = te.eventMatchOptions();
for (dvui.events()) |*e| {
    // for global shortcuts, don't call eventMatch and do it early in the frame
    if (!dvui.eventMatch(e, emo))
        continue;

    if (e.evt == .key and e.evt.key.code == .enter and e.evt.key.action == .down) {
        e.handled = true;  // prevent normal processing
        // handle the shortcut here
    }

    if (!e.handled) {
        te.processEvent(e, false);
    }
}

try te.draw();
te.deinit();
```

**Comment by iacore** [comment]
Yes this is very good.
